wallet-new/src/Cardano/Wallet/Kernel/CoinSelection/Generic/Fees.hs

```diff
@@ -35,107 +36,167 @@ data FeeOptions dom = FeeOptions {
     , foExpenseRegulation :: ExpenseRegulation
     }
 
+data CoinSelFinalResult dom = CoinSelFinalResult
+    { csrInputs  :: NonEmpty (UtxoEntry dom)
+    -- ^ Picked inputs
+    , csrOutputs :: NonEmpty (Output dom)
+    -- ^ Picked outputs
+    , csrChange  :: [Value dom]
+    -- ^ Resulting changes
+    }
+
+type PickUtxo m utxo
+    =  Value (Dom utxo)
+    -> CoinSelT utxo CoinSelHardErr m (Maybe (UtxoEntry (Dom utxo)))
+
 -- | Given the coin selection result from a policy run, adjust the outputs
 -- for fees, potentially returning additional inputs that we need to cover
 -- all fees.
-adjustForFees :: forall utxo m. (CoinSelDom (Dom utxo), Monad m)
-              => FeeOptions (Dom utxo)
-              -> (Value (Dom utxo) ->
-                   CoinSelT utxo CoinSelHardErr m (Maybe (UtxoEntry (Dom utxo))))
-              -> [CoinSelResult (Dom utxo)]
-              -> CoinSelT utxo CoinSelHardErr m
-                   ([CoinSelResult (Dom utxo)], SelectedUtxo (Dom utxo), Value (Dom utxo))
+adjustForFees
+    :: forall utxo m. (CoinSelDom (Dom utxo), Monad m)
+    => FeeOptions (Dom utxo)
+    -> PickUtxo m utxo
+    -> [CoinSelResult (Dom utxo)]
+    -> CoinSelT utxo CoinSelHardErr m (CoinSelFinalResult (Dom utxo))
 adjustForFees feeOptions pickUtxo css = do
-    case foExpenseRegulation feeOptions of
-      ReceiverPaysFee -> coinSelLiftExcept $
-        (, emptySelection, valueZero) <$> receiverPaysFee upperBound css
-      SenderPaysFee ->
-        senderPaysFee pickUtxo upperBound css
-  where
-    upperBound = feeUpperBound feeOptions css
+    let inps = concatMap (selectedEntries . coinSelInputs) css
+    let outs = map coinSelOutput css
+    let chgs = concatMap coinSelChange css
+
+    (inps', outs', chgs') <-
+        case foExpenseRegulation feeOptions of
+          ReceiverPaysFee ->
+            coinSelLiftExcept $ receiverPaysFee feeOptions inps outs chgs
+
+          SenderPaysFee ->
+            senderPaysFee pickUtxo feeOptions inps outs chgs
+
+    let neInps = case inps' of
+            []   -> error "adjustForFees: empty list of inputs"
+            i:is -> i :| is
+
+    let neOuts = case outs' of
+            []   -> error "adjustForFees: empty list of outputs"
+            o:os -> o :| os
+
+    return $ CoinSelFinalResult neInps neOuts chgs'
```

Comment:
@KtorZ We should filter out dust in change here. I found a case where we create a tx with an output of value 0. Is this valid protocol wise :thinking: ? The tx failed, not sure if it's because of this or fees underestimate.

Reply:
Ho dang. You're right, I forgot about re-applying the `dust` filtering. Was a `TODO` in my head.